### ISSUE TITLE
[Headless MFA Deactivation] Always require reauthentication

### ISSUE THREAD

**Issue by mecampbellsoup** [pennersr/django-allauth#4017]
For sensitive auth-related actions like [deactivating one's 2FA](https://docs.allauth.org/en/latest/headless/openapi-specification/#tag/Account:-2FA/paths/~1_allauth~1%7Bclient%7D~1v1~1account~1authenticators~1totp/delete), we would ideally always require the user to reauthenticate. Right now it seems like that is a time-based check:

> Deactivates TOTP authentication. If the user authentication is not sufficiently recent, a reauthentication flow (401) will is presented.

Questions:

1. Would setting the auth timeout value to 0 suit our needs, or would it make other aspects of using the app too cumbersome?
2. Would you be open to adding configuration such that we can say "for deactivating MFA, we _specifically_ require re-authentication _via MFA_"?
3. Specific to deactivating a 2nd factor, would you be open to making an OTP code a required input? I feel like that's more standard UX - you have to enter an OTP from your 2FA device and submit that to disable that 2nd factor.

**Comment by pennersr** [comment]
1. Doubt so. It will likely result in in being asked to reauthenticate over and over again.
2. Not sure. In light of having multiple authenticators, it will result in inconsistencies. For example, you typically don't want to give hardware security keys the same treatment. Because, if I lost one of my security keys, I wouldn't ever be able to get rid of it if you force to reauthenticate using that key.
3. Similar to point 2. 

> I feel like that's more standard UX

Overall, I feel we need more convincing arguments/analysis for changing the existing behavior. For example, I believe GitHub allows you to reauthenticate using any means available.

**Comment by mecampbellsoup** [comment]
> For example, I believe GitHub allows you to reauthenticate using any means available.

I also tend to start by asking "how does GitHub do this?", and to that end, I found a good summary of GitHub's UX [here](https://security.stackexchange.com/a/154934/132533).

I think https://github.com/pennersr/django-allauth/issues/4019 might be related so I'll reference it here, but I'm starting to feel like having a way (i.e. API endpoint) to answer "does this user need to reauthenticate to perform action `foo`?" would be helpful in building our UI/UX however we desire.

In the linked example, I imagine GH is prompting for the _password_ specifically because their backend recognizes the other factor in a disable TOTP 2FA operation is the user's password, but I may be wrong about that. I can't imagine GitHub expects you to possess the factor you're trying to reset though.

Along those lines, I tried going into GitHub (where I have 2FA enabled on my account) and changing my password. If you know your old password and are logged in, it seems they do NOT prompt for an OTP from my 2nd factor. (It seems like they consider correctly entering your old password sufficient authorization which makes sense I guess.)

However, when I am logged in but cannot produce my old password (click `Forgot password?`), I get an email sent to me, so I guess ownership of an email address is considered sufficient proof to authorize sensitive changes like resetting a password.

![image](https://github.com/user-attachments/assets/fc2a4646-a8c2-4965-9e7c-fd056c307538)
